partiql-planner/src/main/kotlin/org/partiql/planner/PartiQLPlannerDefault.kt

```diff
@@ -34,17 +35,28 @@ internal class PartiQLPlannerDefault(
 
         // 3. Resolve variables
         val typer = PlanTyper(env, onProblem)
+        var planStmt = typer.resolve(root)
+
+        // 4. Fold and re-resolve
+        planStmt = fold(planStmt, env, onProblem)
+
         var plan = partiQLPlan(
             version = version,
             globals = env.globals,
-            statement = typer.resolve(root),
+            statement = planStmt,
         )
 
-        // 4. Apply all passes
+        // 5. Apply all passes
         for (pass in passes) {
             plan = pass.apply(plan, onProblem)
         }
 
         return PartiQLPlanner.Result(plan, emptyList())
     }
+
+    private fun fold(stmt: org.partiql.plan.Statement, env: Env, onProblem: ProblemCallback): org.partiql.plan.Statement {
+        val foldedPlan = ConstantFolder.fold(stmt)
+        val typer = PlanTyper(env, onProblem)
+        return typer.resolve(foldedPlan)
```

Comment:
This might be expensive, but we aren't at the point of optimizing yet. Why do we need another typing pass?

Reply:
Unfortunately, we don't have the ability to fold an expression and tell all of it's dependents. Take the `SELECT *` transformation for example:
```
SELECT VALUE TUPLEUNION(
  CASE
    WHEN t IS STRUCT THEN t
    ELSE { '_1': t }
  END
)
FROM <<
  { 'a': 1}
>> AS t
```

When we originally type this, the StaticType of `Rex.Op.Select` will be `BAG(UNION(STRUCT(a), STRUCT(_1)))`. However, once it passes through the constant folder, we drop the second branch since the first branch's condition will always be true. Since the type of `Rex.Op.Select` is dependent upon the type of the CASE expression, we need to retype the plan with the second branch removed.

Reply:
That being said, this also reinforces our curiosity in typing during the conversion from AST to Plan. If a user explicitly writes a `CASE` with multiple potential output types, then I'd assume that, similar to Kotlin, the `CASE`'s output would also be multiple output types (even if we folded out branches that are never to execute). However, if we normalize a `SELECT *` to internally use a `STRUCT` or `TUPLEUNION`, then we can sidestep saying it could be multiple output types. We can definitively say it will be a single type.

Reply:
> When we originally type this, the StaticType of Rex.Op.Select will be BAG(UNION(STRUCT(a), STRUCT(_1))). However, once it passes through the constant folder, we drop the second branch since the first branch's condition will always be true.

Why do we not know that `t` is a struct when we initially type? I think it's naive to type to

 `<< ( closed::{'a': int}, closed::{_1: any} >>`

What I think you should do is _type each branch_, prune branches, then union the remaining branch types. 

I'm writing this on github so I'm making some assumptions here

```kotlin
fun visitRexOpCase(op: Rex.Op.Case, ctx: StaticType): Rex {
   val branches: List<Rex.Op.Case.Branch> = op.branches.map { // type the branch cond+rex }.prune()
   return when (branches.size) {
      0 -> { // err always missing, no valid branch }
      1  -> branches.first.rex // drop the case-when
     else -> {
         // rewrite the case-when with new list of branches
         val type = unionOf(branches.map { it.rex.type })
         val op = rexOpCase(branches)
         rex(type, op)
     }
   }
}

fun List<Rex.Op.Case.Branch>.prune(): List<Rex.Op.Case.Branch> {
    val newBranches = mutableList<Rex.Op.Case.Branch>()
    // apply logic very similar to your constant folding
    // ie cond == true, then short circuit returning a singleton list of this branch
    // cond == false, drop this branch!
}
```